Api endpoints for user with enviroment

HarlamovBuldog opened the pull request.

HarlamovBuldog committed:
Add service interface and models

Service interface implements all methods for working directly with database.

HarlamovBuldog committed:
Add package "storage" to work with db directly

DB struct type holds connection to database.
All functions are methods of DB struct itself.
Tournament.go contains only template for now, just to implement Service type from sts package.

HarlamovBuldog committed:
Add server package for running web server itself

Server struct implements http.Handler and also has sts.Service variable.
Add all entrypoints for user according to technical task.

HarlamovBuldog committed:
Add entrypoint for webserver with config file

pisarukv commented on config.yaml:
```diff
@@ -0,0 +1,3 @@
+connstr: mongodb://localhost:27017
+port: :8000
+dbname: sts
```
no new line at end of file

pisarukv commented on main.go:
```diff
@@ -0,0 +1,61 @@
+)
+
+type config struct {
+	ConnStr string `yaml:"connstr"`
```
why str? isn't obvious that conn is a string?

pisarukv replied:
use snake case in yaml ex. conn_str

pisarukv commented on main.go:
```diff
@@ -0,0 +1,61 @@
+func main() {
+	yamlConfigFile, err := ioutil.ReadFile("config.yaml")
+	if err != nil {
+		log.Printf("error opening cofiguration yaml file: %v\n", err)
```
why not to use log.Fatalf instead combination of printf + return

pisarukv commented on main.go:
```diff
@@ -0,0 +1,61 @@
+	}
+	err = client.Ping(context.TODO(), nil)
+	if err != nil {
+		log.Printf("error connecting to mongo db: %v", err)
```
I belief you missed a return

pisarukv commented on main.go:
```diff
@@ -0,0 +1,61 @@
+
+	log.Println("Connected to MongoDB!")
+
+	db := storage.CreateNew(client.Database("sts"))
```
you have in config db name. why inline str here?

pisarukv commented on main.go:
```diff
@@ -0,0 +1,61 @@
+
+	db := storage.CreateNew(client.Database("sts"))
+	s := server.NewServer(db)
+	if conf.Port == "" {
```
so, why u checking port in the config not after parsing it? I would better create a method for the config struct, let's name it validate, inside that method check connStr and port for an empty string.

pisarukv commented on pkg/server/server.go:
```diff
@@ -0,0 +1,153 @@
+	service sts.Service
+}
+
+// NewServer constructs a Server, decodes yaml configuration file
```
decodes yaml? assign values? where?

pisarukv commented on pkg/server/server.go:
```diff
@@ -0,0 +1,153 @@
+	"log"
+	"net/http"
+
+	"github.com/HarlamovBuldog/social-tournament-service/pkg/sts"
```
we need to discuss sts package

pisarukv commented on pkg/server/server.go:
```diff
@@ -0,0 +1,153 @@
+	userID, ok := vars["id"]
+	if !ok {
+		w.WriteHeader(http.StatusBadRequest)
+		log.Println("error getting user info: user id is not provided")
```
not sure you need `error getting user info:` here

pisarukv commented on pkg/server/server.go:
```diff
@@ -0,0 +1,153 @@
+		return
+	}
+	enc := json.NewEncoder(w)
+	enc.SetIndent("", " ")
```
comment? or why you need it

HarlamovBuldog replied:
to improve json readability i guess

pisarukv replied:
but why do you want to do it on the api side?

pisarukv commented on pkg/server/server.go:
```diff
@@ -0,0 +1,153 @@
+	enc := json.NewEncoder(w)
+	enc.SetIndent("", " ")
+	err = enc.Encode(&userData)
+	if err != nil {
```
let's inline if and enc.Encode

pisarukv commented on pkg/server/server.go:
```diff
@@ -0,0 +1,153 @@
+		return
+	}
+
+	vars := mux.Vars(req)
```
it's not really principal, but I would better check req path parameters at the start of the method

pisarukv commented on pkg/server/server.go:
```diff
@@ -0,0 +1,153 @@
+}
+
+func (s *Server) addUserBonusPoints(w http.ResponseWriter, req *http.Request) {
+	points := struct {
```
don't you want to move that struct outside the function to the pkg lvl?

pisarukv commented on pkg/storage/user.go:
```diff
@@ -0,0 +1,109 @@
+func (db *DB) AddUser(ctx context.Context, name string) (string, error) {
+	insertResult, err := db.conn.Collection(usersCollectionName).InsertOne(ctx, sts.User{
+		Name:    name,
+		Balance: 0,
```
since balance is a float64 0 will be a default value, so that 0 assignment is not needed

pisarukv commented on pkg/storage/user.go:
```diff
@@ -0,0 +1,109 @@
+		Balance: 0,
+	})
+	if err != nil {
+		return "", fmt.Errorf("error insert doc to collection: %v", err)
```
there is such a great package called https://godoc.org/github.com/pkg/errors I insist on using it's errors.Wrap instead fmt.Errorf

pisarukv replied:
not only here, but overall

pisarukv commented on pkg/storage/user.go:
```diff
@@ -0,0 +1,109 @@
+		return nil, fmt.Errorf("error getting doc from collection: %v", err)
+	}
+	var user sts.User
+	err = docReturned.Decode(&user)
```
inline with if

pisarukv commented on pkg/storage/user.go:
```diff
@@ -0,0 +1,109 @@
+		log.Printf("Error on deleting user: %v\n", err)
+		return fmt.Errorf("Error on deleting user: %v", err)
+	}
+	if deleteResult.DeletedCount < 1 {
```
I would better check if deleted != 1

cali4888 commented on pkg/storage/user.go:
```diff
@@ -0,0 +1,109 @@
+	}
+	update := bson.D{
+		{"$inc", bson.D{
+			{"balance", -points},
```
I don't check it, but is looks like you are not subtracting points

pisarukv commented on pkg/sts/sts.go:
```diff
@@ -0,0 +1,53 @@
+package sts
```
overall I don't like general packages with models and interfaces

HarlamovBuldog committed:
Add .golangci.yml config file to manage linters

HarlamovBuldog committed:
Remove package sts. Packages restructuration.

Following best golang code practice models were moved to the same file where they were actually used.
Service interface was also moved to server.go where it's supposed to be.
Sts package is no longer needed.

HarlamovBuldog committed:
Add errors wrapping in user.go. Code refactoring.

All occurences of fmt.Errorf(...) were changed to errors.Wrap(...).
Now we are using "github.com/pkg/errors" library to handle all errors.
Added comments for imported fucntions.
Logic near ModifiedCount and DeletedCount was reworked.

HarlamovBuldog committed:
Error handling and config rework. Add validate method.

log.Printf with return were changed to log.Fatalf.
Same functionality but shorter.
Add Validate method for config struct.
Reason is to check all fields right after successful parsing config file.
Config struct fields and config.yml fields were renamed for better understanding.
defer func was created for handling mongodb client.Disconnect() errors.

pisarukv commented on main.go:
```diff
@@ -0,0 +1,81 @@
+
+type config struct {
+	ConnStr    string `yaml:"conn_str"`
+	ServerPort string `yaml:"server_port"`
```
why server port is str? maybe int?

HarlamovBuldog replied:
i've validate method where i check if server port is uint. I can do it here and set field type uint, but then i will get not very informative err message from decode.

pisarukv replied:
I would better use uint and use error.wrap to describe an error

pisarukv commented on main.go:
```diff
@@ -0,0 +1,81 @@
+	}
+
+	defer func() {
+		err := client.Disconnect(context.TODO())
```
maybe use here context timeout?

pisarukv commented on main.go:
```diff
@@ -0,0 +1,81 @@
+		}
+	}()
+
+	err = client.Ping(context.TODO(), nil)
```
context timeout?

pisarukv commented on main.go:
```diff
@@ -0,0 +1,81 @@
+	db := storage.CreateNew(client.Database(conf.DBName))
+	s := server.NewServer(db)
+
+	log.Fatal(http.ListenAndServe("localhost:"+conf.ServerPort, s))
```
BTW. You will never get into mongo defer because of log.Fatal. Fatal calls os.exit instead of panic

pisarukv replied:
I would better implement here graceful shutdown

pisarukv commented on pkg/server/server.go:
```diff
@@ -0,0 +1,177 @@
+)
+
+// Service is the wrapper for all methods working with db.
+type Service interface {
```
why service exists inside server pkg?

HarlamovBuldog committed:
Move Service interface from server to storage package.

HarlamovBuldog committed:
Apply proper naming, inline if in server.go

HarlamovBuldog committed:
Move support structs in handlers to the pkg lvl.

HarlamovBuldog committed:
Implement some methods to work with tournament collection.

HarlamovBuldog committed:
Add routes for tournament and templates for functions realizing them.

HarlamovBuldog committed:
Folder "pkg" was renamed to "internal".

HarlamovBuldog committed:
Add graceful shutdown, change config struct field.

Graceful shutdown feature was added to prevent data leak.
ServerPort field's type was changed from string to uint16.
In Validate() func we check if value of ServerPort field is zero.

HarlamovBuldog committed:
Code refactoring

pisarukv commented on main.go:
```diff
@@ -61,21 +61,50 @@ func main() {
 	defer func() {
-		err := client.Disconnect(context.TODO())
+		ctx, cancel := context.WithTimeout(context.TODO(), time.Second*5)
+		defer cancel()
```
not sure that you need to defer cancel()

pisarukv commented on main.go:
```diff
@@ -61,21 +61,50 @@ func main() {
 	}()
 
-	err = client.Ping(context.TODO(), nil)
+	ctx, cancel := context.WithTimeout(context.TODO(), time.Second*5)
```
use background instead todo

cali4888 commented:
fix defer cancel problems and LGTM

HarlamovBuldog committed:
Fix code near context, change field type inside config struct.

Near client.Disconnect() we don't need to call defer cancel().
Cause actually we are shutting down program and we don't care about memory leak in this case.
Near cleint.Ping() we need to change parent context to context.Background() in context.WithTimeout() call.
Cause context.Background() is still alive at that time and we should work with it.
We changed field type for ServerPort inside config struct from uint16 to int64 just to simplify things.
Change Validate func logic for server port. It checks server port range now.
